File: DriversRoutes/Pages/Maps/Controls/BlazorMap.razor.cs
using DataBase.Model.EntitiesRoutes;

using Microsoft.JSInterop;

namespace DriversRoutes.Pages.Maps.Controls
{
    public partial class BlazorMap
    {
        public static Action AfterOnInitializedAsync { get; set; }
        public void Dispose()
        {
            _setCustomerAction -= _setCustomer;
            _setAdvancedMarkerAction -= _setAdvancedMarker;
            _setAdvancedMarkerIconAction -= _setAdvancedMarkerIcon;
            _removeAdvancedMarkerAction -= _removeAdvancedMarker;

            _addDirectionsAction -= _addDirections;
            _removeDirectionsAction -= _removeDirections;

            SetCenter -= SetMyLocation;
            MapDragStart -= StopLisiningLocation;

            _fitMapToAdvancedMarkersAction -= _fitMapToAdvancedMarkers;

            DataBase.Helper.MudBlazorTheme.ActionThemeChanged -= SetIconTheme;
            RemoveListener("dragstart");
        }

        public BlazorMap()
        {
            _setCustomerAction += _setCustomer;
            _setAdvancedMarkerAction += _setAdvancedMarker;
            _setAdvancedMarkerIconAction += _setAdvancedMarkerIcon;
            _removeAdvancedMarkerAction += _removeAdvancedMarker;

            _addDirectionsAction += _addDirections;
            _removeDirectionsAction += _removeDirections;

            SetCenter += SetMyLocation;
            MapDragStart += StopLisiningLocation;

            _fitMapToAdvancedMarkersAction += _fitMapToAdvancedMarkers;


            DataBase.Helper.MudBlazorTheme.ActionThemeChanged += SetIconTheme;
            DataBase.Helper.MudBlazorTheme.SetCurrentTheme(AppInfo.Current.RequestedTheme);
            DataBase.Helper.MudBlazorTheme.ActionThemeChanged?.Invoke();

        }



        private ValueTask InitMat(string mapId, double lat, double lng, CancellationToken token = default)
        {
            return _module.InvokeVoidAsync("initMap", token, mapId, lat.ToString(dot), lng.ToString(dot));
        }
        private ValueTask CalculateRoute(string start, string end, CancellationToken token = default)
        {
            return _module.InvokeVoidAsync("calcRoute", token, start, end);
        }
        private ValueTask SetMapCenter(double lat, double lng, CancellationToken token = default)
        {
            return _module.InvokeVoidAsync("changeCenter", token, lat.ToString(dot), lng.ToString(dot));
        }
        private ValueTask AddListener(string listener, string methodName, string assemblyName, CancellationToken token = default)
        {
            return _module.InvokeVoidAsync("addListener", token, listener, methodName, assemblyName);
        }
        private ValueTask RemoveListener(string listener, CancellationToken token = default)
        {
            return _module.InvokeVoidAsync("removeListener", token, listener);
        }
        private ValueTask AddAdvancedMarker(string id, double lat, double lng, string userIcon, CancellationToken token = default)
        {
            return _module.InvokeVoidAsync("addAdvancedMarker", token, id, lat.ToString(dot), lng.ToString(dot), userIcon);
        }
        private ValueTask RemoveAdvancedMarker(string id, CancellationToken token = default)
        {
            return _module.InvokeVoidAsync("removeAdvancedMarker", token, id);
        }
        private ValueTask SetIconAdvancedMarker(string id, string userIcon, CancellationToken token = default)
        {
            return _module.InvokeVoidAsync("setIconAdvancedMarkerSvg", token, id, userIcon);
        }
        private ValueTask SetPositonAdvancedMarker(string id, double lat, double lng, CancellationToken token = default)
        {
            return _module.InvokeVoidAsync("setPositionAdvancedMarker", token, id, lat.ToString(dot), lng.ToString(dot));
        }
        private ValueTask RemoveCalculateRoute(CancellationToken token = default)
        {
            return _module.InvokeVoidAsync("clearRoute", token);
        }
        private ValueTask FitMapToAdvancedMarkers(CancellationToken token = default)
        {
            return _module.InvokeVoidAsync("fitMapToMarkers", token);
        }


        System.Globalization.NumberFormatInfo dot = new System.Globalization.NumberFormatInfo()
        {
            NumberDecimalSeparator = ".",
        };

        #region CustomerRoutesAction

        private CustomerRoutes _customer;

        private static event Action<CustomerRoutes> _setCustomerAction;
        private void _setCustomer(CustomerRoutes customer)
        {
            _customer = customer;
        }
        public static void OnSetCustomer(CustomerRoutes customer = null)
        {
            _setCustomerAction?.Invoke(customer);
        }

        private static event Action<CustomerRoutes, string> _setAdvancedMarkerAction;
        private void _setAdvancedMarker(CustomerRoutes customer = null, string text = "")
        {
            if (customer is null)
            {
                if (_customer is null)
                {
                    return;
                }
                customer = _customer;
            }

            var lat = customer.Latitude;
            var lng = customer.Longitude;
            var img = Data.DrawIconOnMap.CreatePinSvg(text);
            AddAdvancedMarker(customer.Id.ToString(), lat, lng, img);

        }
        /// <summary>
        /// Dodaje advanced marker z dana pozycją
        /// </summary>
        /// <param name="customer">Jeżeli customer nie zostanie ustawiony zostanie dodany marker z prywatnego pola</param>
        public static void OnSetAdvancedMarker(CustomerRoutes customer = null, string text = "")
        {
            _setAdvancedMarkerAction?.Invoke(customer, text);
        }

        private static event Action<CustomerRoutes> _removeAdvancedMarkerAction;
        private void _removeAdvancedMarker(CustomerRoutes customer = null)
        {
            if (customer is null)
            {
                if (_customer is null)
                {
                    return;
                }
                customer = _customer;
            }
            RemoveAdvancedMarker(customer.Id.ToString());
        }
        /// <summary>
        /// Usuwa Advanced marker
        /// </summary>
        /// <param name="customer">Jeżeli customer nie zostanie ustawiony zostanie wybrany z prywatnego pola</param>
        public static void OnRemoveAdvancedMarker(CustomerRoutes customer = null)
        {
            _removeAdvancedMarkerAction?.Invoke(customer);
        }


        private static event Action<CustomerRoutes, string> _setAdvancedMarkerIconAction;
        private void _setAdvancedMarkerIcon(CustomerRoutes customer = null, string icon = "")
        {
            if (customer is null)
            {
                if (_customer is null)
                {
                    return;
                }
                customer = _customer;
            }
            SetIconAdvancedMarker(customer.Id.ToString(), icon);
        }
        /// <summary>
        /// Zmienia icone advance marker
        /// </summary>
        /// <param name="customer">Jeżeli customer nie zostanie ustawiony zostanie ustawiony z prywatnego pola</param>
        /// <param name="icon">icona jako svg</param>
        public static void OnSetAdvancedMarkerIcon(CustomerRoutes customer = null, string icon = "")
        {
            _setAdvancedMarkerIconAction?.Invoke(customer, icon);
        }




        private static event Func<CustomerRoutes, CancellationToken, Task> _addDirectionsAction;
        private async Task _addDirections(CustomerRoutes customer = null, CancellationToken token = default)
        {
            if (customer is null)
            {
                if (_customer is null)
                {
                    return;
                }
                customer = _customer;
            }

            var currentLocation = await Helper.CurrentLocation.Get(GeolocationAccuracy.Best, TimeSpan.FromSeconds(1));

            var latLngStart = $"{currentLocation.Center.Latitude.ToString(dot)},{currentLocation.Center.Longitude.ToString(dot)}";

            var latLngEnd = $"{customer.Latitude.ToString(dot)},{customer.Longitude.ToString(dot)}";

            await CalculateRoute(latLngStart, latLngEnd, token);

        }
        /// <summary>
        /// Dodaje trasę z twojej aktualnej pozycji do wybranego puntku
        /// </summary>
        /// <param name="customer">Jeżeli customer nie zostanie ustawiony zostanie wybrany puntk z prywatnego pola</param>
        /// <returns></returns>
        public static async Task OnAddDirections(CustomerRoutes customer = null, CancellationToken token = default)
        {
            await _addDirectionsAction?.Invoke(customer, token);
        }

        private static event Action<CancellationToken> _removeDirectionsAction;
        private void _removeDirections(CancellationToken token = default)
        {
            RemoveCalculateRoute(token);
        }
        public static void OnRemoveDrirections(CancellationToken token = default)
        {
            _removeDirectionsAction?.Invoke(token);
        }

        private static event Action<CancellationToken> _fitMapToAdvancedMarkersAction;
        private void _fitMapToAdvancedMarkers(CancellationToken token = default)
        {
            FitMapToAdvancedMarkers(token);
        }
        public static void OnFitMapToAdvancedMarkers(CancellationToken token = default)
        {
            _fitMapToAdvancedMarkersAction?.Invoke(token);
        }


        #endregion




    }
}

File: DriversRoutes/Pages/Maps/Controls/BlazorMap.razor.js
let map, directionsService, directionsRenderer;
let valueListener = [];
let valueAdvancedMarkers = [];

export async function initMap(elementId, lat, lng) {
    try {

        const { Map } = await google.maps.importLibrary("maps");
        const { AdvancedMarkerElement } = await google.maps.importLibrary("marker");

        let center = new google.maps.LatLng(lat, lng);
        let mapOptions = {
            zoom: 15,
            center: center,
            mapId: "mapId",
        };

        map = new google.maps.Map(document.getElementById(elementId), mapOptions);
        directionsService = new google.maps.DirectionsService();
        directionsRenderer = new google.maps.DirectionsRenderer();
        directionsRenderer.setMap(map);


    } catch (e) {
        console.error("Error when map was init ", e);
    }


}
export function calcRoute(start, end) {
    try {
        var request = {
            origin: start,
            destination: end,
            travelMode: 'DRIVING',
            language: "pl"
        };

        directionsService.route(request, function (result, status) {
            if (status == 'OK') {
                directionsRenderer.setDirections(result);
                console.debug("Route Direction is set");
            }
        });
    } catch (e) {
        console.error("Error when calculate route " + e);
    }

}
export function clearRoute() {
    if (directionsRenderer) {
        directionsRenderer.setDirections({ routes: [] });
    }
}


export function changeCenter(lat, lng) {
    try {
        let newCenter = new google.maps.LatLng(lat, lng);
        map.setCenter(newCenter);
        console.debug("Map set center " + lat + " " + lng);
    } catch (e) {
        console.error("Error when map set center " + e);
    }

}
export function addListener(lisinerName, methodName, assemblyName) {
    try {
        let listener = map.addListener(lisinerName, function () {
            DotNet.invokeMethodAsync(assemblyName, methodName);
            console.debug("Listener invoke " + lisinerName);
        });
        valueListener.push(listener);
        console.debug(listener);
        console.debug("Listener add to list " + lisinerName);
    } catch (e) {
        console.error("Error when listener was add to map " + e);
    }

}
export function removeListener(listenerName) {
    try {
        let index = findListener(valueListener, listenerName)
        if (index == -1) {
            console.debug("Listener not found")
            return;
        }
        google.maps.event.removeListener(valueListener[index]);
        console.debug("Listener remove  from map" + listenerName);
        removeItemAll(valueListener, index);
        console.debug("Listener remove from list " + listenerName);
    } catch (e) {
        console.error("Error when listener was remove from map" + e);
    }

}

function removeItemAll(arr, index) {
    let i = 0;
    while (i < arr.length) {
        if (i === index) {
            arr.splice(i, 1);
        } else {
            ++i;
        }
    }
    return arr;
}
function findListener(arr, value) {
    for (var i = 0; i < arr.length; i++) {
        if (arr[i].Eg == value) {
            return i;
        }
    }
    return -1;
}


export function addAdvancedMarker(id, lat, lng, userIcon) {

    try {
        let latLng = new google.maps.LatLng(lat, lng);
        let parser = new DOMParser();
        let pinSvg = parser.parseFromString(
            userIcon,
            "image/svg+xml",
        ).documentElement;

        let marker = new google.maps.marker.AdvancedMarkerElement({
            position: latLng,
            map: map,
            content: pinSvg,
        });
        marker.customId = id;
        console.debug("Map add advance marker " + id);
        valueAdvancedMarkers.push(marker);
        console.debug("Map Add advanced marker to list " + id);
    }
    catch (error) {
        console.error("Error when Advanced marker was created:", error);
    }
}
export function removeAdvancedMarker(id) {
    try {
        console.debug(valueAdvancedMarkers);
        let index = findMarker(valueAdvancedMarkers, id);
        let marker = valueAdvancedMarkers[index];
        marker.setMap(null);
        removeItemAll(valueAdvancedMarkers, index);
        console.debug("Advanced marker delete " + id)
    }
    catch (e) {
        console.error("Error when remove adwanced marker " + e);
    }
}
export function setIconAdvancedMarkerSvg(id, userIcon) {
    try {
        let index = findMarker(valueAdvancedMarkers, id);
        let marker = valueAdvancedMarkers[index];

        let parser = new DOMParser();
        let pinSvg = parser.parseFromString(
            userIcon,
            "image/svg+xml",
        ).documentElement;

        marker.content = pinSvg;
        marker.draw();
        console.debug("Advanced marker set new icon " + id)
    }
    catch (e) {
        console.error("Error when Advanced marker set new icon " + e);
    }

}
export function setPositionAdvancedMarker(id, lat, lng) {
    try {
        let index = findMarker(valueAdvancedMarkers, id);
        let marker = valueAdvancedMarkers[index];
        let latLng = new google.maps.LatLng(lat, lng);

        marker.position = latLng,
            marker.draw();
        console.debug("Advanced marker set new position " + id)
    } catch (e) {
        console.error("Error when Advanced marker set new position " + e);
    }

}

function findMarker(arr, value) {
    for (var i = 0; i < arr.length; i++) {
        if (arr[i].customId == value) {
            return i;
        }
    }
    return -1;
}

export function fitMapToMarkers() {
    try {
        let bounds = new google.maps.LatLngBounds();
        valueAdvancedMarkers.forEach(marker => bounds.extend(marker.position));
        map.fitBounds(bounds);
    } catch (e) {
        console.error("Error when bounds pins", e);
    }
}